src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(isIbd), isIbd);
+
+            lock (this.peerLock)
+            {
+                foreach (IBlockPullerBehavior blockPullerBehavior in this.pullerBehaviorsByPeerId.Values)
+                    blockPullerBehavior.OnIbdStateChanged(isIbd);
+
+                this.isIbd = isIbd;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void NewPeerTipClaimed(INetworkPeer peer, ChainedHeader newTip)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(peer.Connection.Id), peer.Connection.Id, nameof(newTip), newTip);
+
+            lock (this.peerLock)
+            {
+                int peerId = peer.Connection.Id;
+
+                if (this.pullerBehaviorsByPeerId.ContainsKey(peerId))
+                {
+                    this.pullerBehaviorsByPeerId[peerId].Tip = newTip;
+                    this.logger.LogTrace("Tip for peer with ID {0} was changed to '{1}'.", peerId, newTip);
+                }
+                else
+                {
+                    bool supportsRequirments = this.networkPeerRequirement.Check(peer.PeerVersion);
+
+                    if (supportsRequirments)
+                    {
+                        var behavior = peer.Behavior<IBlockPullerBehavior>();
+                        behavior.Tip = newTip;
+                        this.pullerBehaviorsByPeerId.Add(peerId, behavior);
+
+                        this.logger.LogTrace("New peer with ID {0} and tip '{1}' was added.", peerId, newTip);
+                    }
+                    else
+                        this.logger.LogTrace("Peer ID {0} was discarded since he doesn't support the requirements.", peerId);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void PeerDisconnected(int peerId)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(peerId), peerId);
+
+            lock (this.peerLock)
+            {
+                this.pullerBehaviorsByPeerId.Remove(peerId);
+            }
+
+            this.ReleaseAndReassignAssignments(peerId);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void RequestBlocksDownload(List<ChainedHeader> headers)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(headers.Count), headers.Count);
+
+            lock (this.queueLock)
+            {
+                // Enqueue new download job.
+                int jobId = this.nextJobId++;
+
+                this.downloadJobsQueue.Enqueue(new DownloadJob()
+                {
+                    Headers = new List<ChainedHeader>(headers),
+                    Id = jobId
+                });
+
+                this.logger.LogDebug("{0} blocks were requested from puller. Job ID {1} was created.", headers.Count, jobId);
+                this.processQueuesSignal.Set();
```

Comment:
Should this happen inside the lock? `processQueuesSignal.Set()`

Reply:
Also take the trace out of the lock

Reply:
>Should this happen inside the lock?

it is under the queueLock (taken on 325)

Reply:
>Also take the trace out of the lock

Why? It's cheap to call. Also it logs smth that was acquired inside a lock so if we are to move trace out it will require moving the jobId out of the lock.

In general we log inside the locks in many cases in many components so it should be fine.

Reply:
Yes but we should try to not log in side locks, if we do its because we have to (if its an intentionally a big lock) put the job id outside whats the problem?

Reply:
Ok, makes sense, fixed

Reply:
Much more important than having the log outside of the lock is having the logs in correct order. And here we need to know before the set the signal. So please put it back inside of the lock.